File: src/main/java/com/github/svegon/capi/event/network/PacketSendCallback.java
package com.github.svegon.capi.event.network;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.network.ClientConnection;
import net.minecraft.network.packet.Packet;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

public interface PacketSendCallback {
    Event<PacketSendCallback> EVENT = EventFactory.createArrayBacked(PacketSendCallback.class,
            (connection, packet, callback) -> {}, listeners -> (connection, packet, callback) -> {
        for (PacketSendCallback listener : listeners) {
            listener.onPacketSend(connection, packet, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
            });

    void onPacketSend(ClientConnection connection, Packet<?> packet, CallbackInfo callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/network/PacketSendCallback.kt
package io.github.svegon.capi.event.network

import io.github.svegon.capi.event.network.PacketSendCallback
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.network.ClientConnection
import net.minecraft.network.packet.Packet
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import java.util.function.Function
import kotlin.Array
import kotlin.invoke

interface PacketSendCallback {
    fun onPacketSend(connection: ClientConnection?, packet: Packet<*>?, callback: CallbackInfo?)

    companion object {
        @JvmField
        val EVENT: Event<PacketSendCallback?> = EventFactory.createArrayBacked(
            PacketSendCallback::class.java,
            PacketSendCallback { connection: ClientConnection?, packet: Packet<*>?, callback: CallbackInfo? -> },
            Function { listeners: Array<PacketSendCallback?>? ->
                PacketSendCallback { connection: ClientConnection?, packet: Packet<*>?, callback: CallbackInfo? ->
                    for (listener in listeners) {
                        listener.onPacketSend(connection, packet, callback)

                        if (callback.isCancelled()) {
                            return@PacketSendCallback
                        }
                    }
                }
            })
    }
}
